src/lib.rs

```diff
@@ -0,0 +1,353 @@
+impl Debug for Object {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Object {{ addr: 0x{:x}, name: {:?}, phdrs: {:?}, num_phdrs: {} }}",
+               self.addr, self.name, self.phdrs, self.num_phdrs)
+    }
+}
+
+pub struct ProgramHeader(*const p_ffi::Elf_Phdr);
+
+impl ProgramHeader {
+    /// Returns the segment type (as one of the `PT_*` constants).
+    pub fn typ(&self) -> Elf_Word {
+        (unsafe{*self.0}).p_type
+    }
+
+    /// Returns the segment flags bitfield.
+    /// See the `PT_*` constants for the meaning of the bitfield.
+    pub fn flags(&self) -> Elf_Word {
+        (unsafe{*self.0}).p_flags
+    }
+
+    /// Returns the segment offset.
+    pub fn offset(&self) -> Elf_Off {
+        (unsafe{*self.0}).p_offset
+    }
+
+    /// Returns the segment virtual address.
+    pub fn vaddr(&self) -> Elf_Addr {
+        (unsafe{*self.0}).p_vaddr
+    }
+
+    /// Returns the segment physical address.
+    /// On modern systems, this is usually reported as the same as the virtual address.
+    pub fn paddr(&self) -> Elf_Addr {
+        (unsafe{*self.0}).p_paddr
+    }
+
+    /// Returns the size of the segment when on disk.
+    pub fn filesz(&self) -> Elf_Xword {
+        (unsafe{*self.0}).p_filesz
+    }
+
+    /// Returns the size of the segment when in memory.
+    pub fn memsz(&self) -> Elf_Xword {
+        (unsafe{*self.0}).p_memsz
+    }
+
+    /// Returns the alignment of the segment.
+    pub fn align(&self) -> Elf_Xword {
+        (unsafe{*self.0}).p_align
+    }
+}
+
+impl Debug for ProgramHeader {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut to_write = String::from("ProgramHeader(");
+
+        let typ = self.typ();
+        let typ_str = match typ {
+            PT_NULL => "PT_NULL",
+            PT_LOAD => "PT_LOAD",
+            PT_DYNAMIC => "PT_DYNAMIC",
+            PT_INTERP => "PT_INTERP",
+            PT_NOTE => "PT_NOTE",
+            PT_SHLIB => "PT_SHLIB",
+            PT_PHDR => "PT_PHDR",
+            PT_TLS => "PT_TLS",
+            PT_LOOS => "PT_LOOS",
+            PT_HIOS => "PT_HIOS",
+            PT_LOPROC => "PT_LOPROC",
+            PT_HIPROC => "PT_HIPROC",
+            PT_GNU_EH_FRAME => "PT_GNU_EH_FRAME",
+            PT_GNU_RELRO => "PT_GNU_RELRO",
+            // Some operating systems define their own non-standard segment types.
+            _ => "other",
+        };
+        to_write.push_str(&format!("typ={} ({}), ", typ, typ_str));
+
+        let flags = self.flags();
+        let mut flag_strs = Vec::new();
+        if flags & PF_X != 0 {
+            flag_strs.push("PF_X");
+        }
+        if flags & PF_W != 0 {
+            flag_strs.push("PF_W");
+        }
+        if flags & PF_R != 0 {
+            flag_strs.push("PF_R");
+        }
+        if flags & PF_MASKPROC != 0 {
+            flag_strs.push("PF_MASKPROC");
+        }
+        to_write.push_str(&format!("flags=<{}>, ", flag_strs.join("|")));
+
```

Comment:
Slightly random blank line?

Reply:
The last 15 lines all deal with one field. It separates this from the following fields, which have one LoC per field.

Reply:
OK.